allennlp/common/from_params.py

```diff
@@ -126,6 +126,10 @@ def infer_params(cls: Type[T], constructor: Callable[..., T] = None) -> Dict[str
         if param.kind == param.VAR_KEYWORD:
             has_kwargs = True
 
+    args_param = parameters.get("args", None)
```

Comment:
Hmm, the point of line 116 above was to handle this case.  Looks like I didn't do it well enough, because it doesn't handle intermediate classes that also have no constructor.

I was trying to decide if this is really what we want to do with a `*args` parameter, and I think it probably is; we don't really have any basis for constructing anything in `*args`, nor do we have a good way of passing in those arguments.

So, yes, I think this is a good change, but can you also do two small other changes, to handle this whole thing more cleanly?  First, remove the check for `FromParams` on line 116 above, and second, add a bare constructor to `FromParams`, with a docstring saying that this exists so that argument inference on super classes hits a bare constructor in the base case, instead of the `object` constructor, which has `*args` and `**kwargs` which we don't want.

Reply:
So just to makes things clear:
- Adding a bare constructor to `FromParams` solves the problem of reverting to `object`'s constructor.
- Keeping the code which explicitly removes `*args` is good as well because we don't use it in `FromParams`, even if it is explicit in the class constructor. For this reason, I remove any VAR_POSITIONAL and not only ones named `args`.

Reply:
@matt-gardner Adding a bare constructor to `FromParams` is causing failures when initializing `Optimizer`s, because it gets to the `FromParams` constructor instead of `torch.optim.*`'s.
Changing the inheritance order fixes this issue, but I am not sure what side effects it might have.
Test-wise, it only causes type check (https://github.com/allenai/allennlp/pull/4629/checks?check_run_id=1095095838) to fail. Didn't check how to fix it.